packages/effect/src/HashMap.ts

```diff
@@ -194,6 +194,27 @@ export const hasHash: {
   <K1 extends K, K, V>(self: HashMap<K, V>, key: K1, hash: number): boolean
 } = HM.hasHash
 
+/**
+ * Checks if an element matching the given predicate exists in the given `HashMap`.
+ *
+ * @example
+ * ```ts
+ * import { HashMap } from "effect"
+ *
+ * const hm = HashMap.make([1, 'a'])
+ * HashMap.hasBy(hm, (value, key) => value === 'a' && key === 1); // -> true
+ * HashMap.hasBy(hm, (value) => value === 'b'); // -> false
+ *
+ * ```
+ *
+ * @since 3.16.0
+ * @category elements
+ */
+export const hasBy: {
+  <K, V>(predicate: (value: V, key: K) => boolean): (self: HashMap<K, V>) => boolean
```

Comment:
You will need NoInfer here, to prevent the predicate from over-riding the types.